### ISSUE TITLE
Update module structure

### ISSUE THREAD

**Issue by sts** [claranet/puppet-consul_template#100]
Adopt module structure to the [Github internal fork](https://github.com/github/puppet-consul_template). It has several advantages:

- The interfaces provide a config_hash and config_defaults instead of individual parameters to both allow access to the full set of configuration options and avoid the need to make module changes as consul-template configuration evolves.

- It allows a much leaner implementation, adding options to consul_template::config is quite confusing.

**Comment by sts** [comment]
@craigwatson any thoughts on the future of this module? I'd like to discuss this before investing any further work.

**Comment by craigwatson** [comment]
@sts Apologies for the delay - my knowledge of Consul itself is admittedly lacking, but I see no downside to adopting the GitHub fork's structure, as you mentioned above it allows for a much cleaner module implementation.

Would you mind creating a PR for getting the fork back in-sync with this repo?

**Comment by wyardley** [comment]
@sts @craigwatson I can try to help on trying to incorporate those changes, since we need those too. However, I think first, it might be a good idea to implement data types and change the class structure around, getting rid of the contain pattern - I'll submit a PR that you can take a look at.
Because of the drift (even before any structural changes), it may be hard to directly cherry-pick the changes.

**Comment by wyardley** [comment]
@ross - it seems like this fork is maintained again, are you still using this at Github, and would you be able / willing to help reconcile some of the changes in the fork?

If not, I will see if @duffrecords can work on adding some features (like the `config_hash` / `config_defaults`) back in.

**Comment by ross** [comment]
Hi @wyardley our fork is still in use, though not actively in development since it was just updated to do the bits we needed and we otherwise hadn't needed anything more from it. I was sad when the original PR on the origin stalled. We'd definitely look to go back to master if this happened, but unfortunately I won't personally have the bandwidth to look into getting our changes back in.

**Comment by sts** [comment]
I prepared a branch with the github module style in my fork. https://github.com/sts/puppet-consul_template/tree/github_module_style

One thing I wasn't sure about yet, is the dependency introduced to the `puppet-consul` module for the `consul_sorted_json` function, which is used to render the hash to the configuration file. Shall we keep it or rather copy it to a `consul_template_sorted_json`?

/cc @wyardley @ross

**Comment by sts** [comment]
@craigwatson Please stop accepting merges for new configuration values, until this is resolved otherwise we need to keep rebasing that patches. People should maintain them in their own forks until we are sorted here.

**Comment by craigwatson** [comment]
@sts thanks very much for the PR, apologies for the mess in the mean-time, I'll hold all PRs until we get the fork mess sorted 👍

**Comment by sts** [comment]
@craigwatson I finished the PR today, please review. I tested the code on Debian already.

**Comment by sts** [comment]
@craigwatson If possible please still delay the forge release, we just discovered a bug with `consul_template::watch::perms`. Its being rendered as an integer value.

**Comment by sts** [comment]
As a workaround one can assign permissions without the leading zero:

```puppet
consul_template::watch { "example_file"
  config_hash => {
    perms => '644'
  }
}
```

**Comment by craigwatson** [comment]
@sts I don't have any plans for a Forge release currently - given the amount of change introduced I'm planning on creating a Vagrant environment for the module to make testing a little easier :)

**Comment by phaer** [comment]
As the closed issues above indicate, the issue with `perms` is now fixed for people using v3.2.3 or later o `puppet-consul`. `de9baa7216ef8ee6f5cdc32aed93d60f1fd7f303` of `puppet-consul_template` works for us too.

**Comment by craigwatson** [comment]
@phaer thanks for the confirmation :)

@sts @wyardley Now that 1.0.0 has been released to the Forge, I'll close this issue 🎉
